### PR TITLE
OpenID Connect authentication

### PR DESCRIPTION
**Pull request by int128**
This pull request allows OpenID Connect authentication. It contains following parts:

- Add the Sign in with OpenID Connect button.
- Add following handlers.
  - POST `/signin/oidc` (initiates an authentication)
  - GET `/signin/oidc` (handles a redirect back from the IdP)
- Add items to the system settings.
- Add the table `ACCOUNT_FEDERATION` to associate a GitBucket user and an external identity.

Please let me know if you have any questions or suggestions.

### How to use

I have tested with Google Identity Provider and Keycloak. I will add below to the Wiki if this PR is merged.

#### Google Identity Provider tutorial:

Setup Google Identity Provider:

1. Open https://console.developers.google.com/apis/credentials and create an OAuth client ID.
1. Application type: web application
1. Authorized redirect URIs: `http://localhost:8080/signin/oidc`
1. Check the client ID and client secret of the created client.

Setup GitBucket:

1. Open the system settings and turn on OpenID Connect.
1. Issuer: `https://accounts.google.com`
1. Client ID: See Google Identity Provider
1. Client secret: See Google Identity Provider
1. Expected signature algorithm: RS256

Note that any Google user can sign in to your GitBucket and make sure restricted people can access to your GitBucket.

See also https://developers.google.com/identity/protocols/OpenIDConnect.

#### Keycloak tutorial:

Setup Keycloak:

1. Create a new Client on your Keycloak.
1. Client ID: `gitbucket`
1. Client Protocol: openid-connect
1. Access Type: confidential
1. Valid Redirect URIs: `http://localhost:8080/signin/oidc`

Setup GitBucket:

1. Open the system settings and turn on OpenID Connect.
1. Issuer: `https://keycloak.example.com/auth/realms/YOUR_REALM` (replace `keycloak.example.com` with your host name and `YOUR_REALM` with your realm)
1. Client ID: `gitbucket`
1. Client secret: See Credetials tab in Keycloak
1. Expected signature algorithm: RS256

### Future work

Followings are not implemented yet:

- Administrator can remove association.
- Existent user can associate an external identity. (currently existent user cannot sign in with OIDC)

### Before submitting a pull-request to GitBucket I have first:

- [x] read the [contribution guidelines](https://github.com/gitbucket/gitbucket/blob/master/.github/CONTRIBUTING.md)
- [x] rebased my branch over master
- [x] verified that project is compiling
- [x] verified that tests are passing
- [x] squashed my commits as appropriate *(keep several commits if it is relevant to understand the PR)*
- [x] [marked as closed using commit message](https://help.github.com/articles/closing-issues-via-commit-messages/) all issue ID that this PR should correct

### CODE REVIEW

**Comment by RadWolfie** [comment]
Would it be wiser to have OpenID as a plugin instead of integrate into gitbucket? That's my opinion.

**Comment by takezoe** [comment]
@RadWolfie Yes, it would be nice if we can add OpenID connect support as a plugin, but it's difficult because the authentication process in GitBucket is not modulable.

I think that OpenID connect support is helpful for users who have already managed their accounts in GoogleApps (including my company :-)). Sometimes it would be an important feature as same as LDAP support. So I want to merge this pull request.

**Comment by RadWolfie** [comment]
@takezoe Understandable, hopefully in the near future we'll have authentication process support for plugins.

File: src/main/scala/gitbucket/core/util/OpenIDConnectUtil.scala [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import scala.collection.JavaConverters.asScalaSet
+
+object OpenIDConnectUtil {
+  val JWS_ALGORITHMS: Map[String, Set[JWSAlgorithm]] = Seq(
+    "HMAC" -> Family.HMAC_SHA,
+    "RSA" -> Family.RSA,
+    "ECDSA" -> Family.EC,
+    "EdDSA" -> Family.ED
+  ).toMap.map { case (name, family) => (name, asScalaSet(family).toSet) }
+}
```
[the remark this answers is not in the document]
Comment (int128) [reply src/main/scala/gitbucket/core/util/OpenIDConnectUtil.scala]: Cool. I will fix it.

File: src/main/scala/gitbucket/core/service/AccountFederationService.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package gitbucket.core.service
+
+import gitbucket.core.model.Profile.profile.blockingApi._
+import gitbucket.core.model.Profile.{AccountFederations, Accounts}
+import gitbucket.core.model.{Account, AccountFederation}
+import gitbucket.core.util.SyntaxSugars.~
+import org.slf4j.LoggerFactory
+
+trait AccountFederationService extends AccountService {
```
[the remark this answers is not in the document]
Comment (int128) [reply src/main/scala/gitbucket/core/service/AccountFederationService.scala]: Okay, I will fix it.

File: src/main/scala/gitbucket/core/controller/IndexController.scala [hunk 3]
```diff
@@ -55,13 +70,61 @@ trait IndexControllerBase extends ControllerBase {
...
+  }
+
+  /**
+    * Handle an OpenID Connect authentication response.
+    */
+  get("/signin/oidc") {
+    context.settings.oidc.map { oidc =>
+      val redirectURI = new URI(s"$baseUrl/signin/oidc")
+      Seq(Keys.Session.OidcState, Keys.Session.OidcNonce, Keys.Session.OidcRedirectBackURI).map(session.get(_)) match {
+        case Seq(Some(state: State), Some(nonce: Nonce), Some(redirectBackURI: String)) =>
```
[the remark this answers is not in the document]
Comment (int128) [reply src/main/scala/gitbucket/core/controller/IndexController.scala]: Good. I will fix it.

**Comment by int128** [comment]
I made some refactors and added a test. Would you check them again? Please let me know if you have more suggestions.

**Comment by takezoe** [comment]
@int128 Thanks! I think the instruction described above should be copied to Wiki as same as LDAP authentication settings. Can I do it?

**Comment by int128** [comment]
@takezoe Thanks, please create a Wiki page.

**Comment by takezoe** [comment]
Created: https://github.com/gitbucket/gitbucket/wiki/OpenID-Connect-Settings

**Pull request merged**